Implement named arguments for NEW DTO syntax

Zuruuh opened the pull request:
I love using DTOs but it can get pretty annoying to understand what's going on when using a ton of arguments, having to keep track of what is what, in what order, etc... Using named arguments should really help a lot with that I think :+1:  
Small note but I also made it that trailing commas don't crash the lexer which is a nice qol feature to have imo. I can remove it / move it to another PR if needed 👌

Now the following DQL works as expected
```sql
SELECT NEW CustomerDTO(
  email: e.email,
  name: u.name,
  city: a.address,
) FROM ...
```

Closes #9182

derrabus commented:
Thank you. Your new tests are failing, however.

Zuruuh commented:
yeah you're right, I only tested locally with a php version above 8 (and it worked fine), but it seems like `ReflectionClass::newInstanceArgs` didn't work with an associative array before the introduction of named params in php, so it will require a bit more work to handle php 7.x versions

greg0ire commented on tests/Doctrine/Tests/ORM/Functional/NewOperatorTest.php:
```diff
@@ -1076,69 +1076,85 @@ public function testClassCantBeInstantiatedException(): void
+                ORDER BY
+                    u.name',
+            ],
+            'Both named and unnamed arguments without trailling comma' => [
```
```suggestion
            'Both named and unnamed arguments without trailing comma' => [
```

greg0ire commented on tests/Doctrine/Tests/ORM/Functional/NewOperatorTest.php:
```diff
@@ -1076,69 +1076,85 @@ public function testClassCantBeInstantiatedException(): void
     }
 
-    public function testQueryWithOnlyNamedArgumentsInOrder(): void
+    /** @return array<string, string> */
```
```suggestion
    /** @return array<string, array{string}> */
```

Zuruuh commented:
Tested locally with php 7.1, tests should pass fine now :+1:

greg0ire commented:
This probably needs to be updated as well:

https://github.com/doctrine/orm/blob/2b91edc525bde4f23f9ada22a82c021b5c18c92b/docs/en/reference/dql-doctrine-query-language.rst?plain=1#L1690-L1691

greg0ire commented on lib/Doctrine/ORM/Query/Parser.php:
```diff
@@ -1938,17 +1946,28 @@ public function NewObjectExpression()
      * @return mixed
      */
-    public function NewObjectArg()
+    public function NewObjectArg(bool $namedArgAlreadyParsed = false)
```
The Parser is neither final nor internal. This means this is a breaking change for extending classes. You can obtain the new argument with `func_get_args()` without changing the signature.
```suggestion
    public function NewObjectArg(/* bool $namedArgAlreadyParsed = false */)
```

derrabus commented on docs/en/reference/dql-doctrine-query-language.rst:
```diff
@@ -584,6 +584,16 @@ And then use the ``NEW`` DQL keyword :
 
 Note that you can only pass scalar expressions to the constructor.
 
+The ``NEW`` operator also supports named arguments, similarly to php 8.0 :
```
```suggestion
The ``NEW`` operator also supports named arguments:
```

derrabus commented on docs/en/reference/dql-doctrine-query-language.rst:
```diff
@@ -584,6 +584,16 @@ And then use the ``NEW`` DQL keyword :
+    $query = $em->createQuery('SELECT NEW CustomerDTO(email: e.email, name: c.name, address: a.city) FROM Customer c JOIN c.email e JOIN c.address a');
+    $users = $query->getResult(); // array of CustomerDTO
+
+Note that you cannot pass unnamed arguments after named ones, just like in php.
```
```suggestion
Note that you must not pass ordered arguments after named ones.
```

derrabus commented on lib/Doctrine/ORM/Internal/Hydration/ObjectHydrator.php:
```diff
@@ -565,7 +568,28 @@ protected function hydrateRowData(array $row, array &$result)
+                    $obj = $class->newInstanceArgs($args);
+                } else {
+                    $constructor = $class->getConstructor();
+                    $unnamedArgs = [];
```
```suggestion
                    $orderedArgs = [];
```

derrabus commented on lib/Doctrine/ORM/Query/AST/NamedScalarExpression.php:
```diff
@@ -0,0 +1,28 @@
+    /** @var Node */
+    public $innerExpression;
+
+    /** @var ?string */
```
```suggestion
    /** @var string|null */
```

derrabus commented on lib/Doctrine/ORM/Query/Parser.php:
```diff
@@ -1938,17 +1947,30 @@ public function NewObjectExpression()
-    public function NewObjectArg()
+    public function NewObjectArg(/* bool $namedArgAlreadyParsed = false */)
     {
+        $namedArgAlreadyParsed = func_get_args()[0] ?? false;
```
Do we actually need this parameter? Couldn't `NewObjectExpression()` just check if the kind of argument it receives is supported at its position?

Zuruuh replied:
I think this should work (but idk how to feel abt it): when parsing a `NewObjectExpression`, we parse the current one, consume the tokens, then parse the next one (if there is any) without consuming it, and throw if it's not named while the current one is. It would work without the need of the parameter, but i feel like it's kinda doing too much ? also we're parsing twice all the arguments while it's not really needed (idk if it's really impactant in terms of perf). wdyt ?

Zuruuh replied:
Hello @derrabus, I didn't want to ping earlier since it was holiday season; what do you think about the approach describe above ? did you have something else in mind ?

Zuruuh committed:
Rebase and squash on branch 3.1

Zuruuh commented:
just saw #11208, I will update the mr to target 3.1 :+1:

Zuruuh commented:
Update done
 - dropped the custom implementation for php versions prior to `8.0` since the lib requires `>=8.1`
 - Implemented the changes that would have been breaking in `src/Query/Parser.php` since the class is now finalized

norkunas commented on src/Internal/Hydration/ObjectHydrator.php:
```diff
@@ -554,7 +554,8 @@ protected function hydrateRowData(array $row, array &$result): void
                 $class = $newObject['class'];
                 $args  = $newObject['args'];
-                $obj   = $class->newInstanceArgs($args);
+
+                $obj = $class->newInstanceArgs($args);
```
what's the point of this diff?

Zuruuh replied:
there used to be code before switching target branch, will remove :+1:

Zuruuh committed:
remove whitespace

beberlei commented:
This was taken over by #11575
